File: javaScript/18-4Sum.js
console.log('start');

var fourSum = function(nums, target) {
    if(nums.length<4) return [];

    nums.sort((a,b)=>a-b);
    result = [];
    let sum=0;
    for (let i=0; i<nums.length-3; i++){
        if(nums[i]==nums[i-1]) continue;
        for (let j=i+1;j<nums.length-2;j++){
            let k=j+1;
            let l=nums.length-1;
            while(k<l && l<nums.length){
                sum = nums[i] + nums[j] + nums[k] + nums[l];
                if(sum==target){
                    result.push([nums[i],nums[j],nums[k],nums[l]]);
                    while (nums[l]==nums[l-1]) l--;
                    while (nums[k]==nums[k+1]) k++;
                    k++;
                    l--;
                } else if (sum>target) {
                    l--;
                } else {
                    k++;
                }
            }
            while (nums[j]==nums[j+1]) j++;
        }
        while (nums[i]==nums[i+1]) i++;
    }

    return result;
    
};

let tests = [];
let solutions = [];

/* tests.push([[1,0,-1,0,-2,2],0]);
solutions.push([[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]]);

tests.push([[2,2,2,2,2],8]);
solutions.push([2,2,2,2]); */

tests.push([[-3,-2,-1,0,0,1,2,3],0]);
solutions.push([[-3,-2,2,3],[-3,-1,1,3],[-3,0,0,3],[-3,0,1,2],[-2,-1,0,3],[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]]);

let result ='';
for (let i=0; i<tests.length; i++){
    result = fourSum(tests[i][0],tests[i][1]);
    console.log('Test ' + i,'||',result,solutions[i]);
}

File: 18_4Sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        if len(nums)<4: return []
        nums.sort()
        ans = []
        i=0
        while i < len(nums)-3:
            j=i+1
            k = j+1
            while j < k:
                k = j+1
                l= len(nums)-1
                while l>k:
                    sum = nums[i] + nums[j] + nums[k] + nums[l] - target
                    if sum > 0:
                        l -=1
                    elif sum < 0:
                        k +=1
                    else:
                        if [nums[i],nums[j],nums[k] ,nums[l]] not in ans: ans.append([nums[i],nums[j],nums[k] ,nums[l]])
                        k +=1
                j+=1
                while nums[j-1]==nums[j] and j<k:
                    j+=1
            i +=1
            while nums[i-1]==nums[i] and i<j:
                    i+=1
        return ans

tests = []
solutions = []

tests.append([[1,0,-1,0,-2,2],0])
solutions.append([[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]])

tests.append([[0,0,0,0],0])
solutions.append([0,0,0,0])

for i in range(len(tests)):
    result = Solution().fourSum(tests[i][0],tests[i][1])
    print('Tests ' + str(i),'Result',result,solutions[i])